pkgpanda/__init__.py

```diff
@@ -77,6 +88,65 @@ def stop_all(self):
                 if ex.returncode != 5:
                     raise
 
+    def remove_staged_unit_files(self):
+        """Remove staged unit files created by Systemd.stage_new_units()."""
+        for filename in os.listdir(self.__base_systemd):
+            if filename.endswith(self.new_unit_suffix):
+                os.remove(os.path.join(self.__base_systemd, filename))
+
+    def stage_new_units(self, new_wants_dir):
+        """Prepare new systemd units for activation.
+
+        Unit files targeted by the symlinks in new_wants_dir are copied to a temporary location in the base systemd
+        directory, and the symlinks are rewritten to target the intended final destination of the copied unit files.
+
+        """
+        for unit_name in self.unit_names(new_wants_dir):
+            wants_symlink_path = os.path.join(new_wants_dir, unit_name)
+            package_file_path = os.path.realpath(wants_symlink_path)
+            systemd_file_path = os.path.join(self.__base_systemd, unit_name)
+            tmp_systemd_file_path = systemd_file_path + self.new_unit_suffix
+
+            # Copy the unit file to the systemd directory with a suffix added to the filename.
+            # This file will be moved to systemd_file_path when the new package set is swapped in.
+            shutil.copyfile(package_file_path, tmp_systemd_file_path)
+            shutil.copymode(package_file_path, tmp_systemd_file_path)
+
+            # Rewrite the symlink to point to the copied unit file's destination.
+            # This symlink won't point to the correct file until the copied unit file is moved to its target location
+            # during activate_new_unit_files().
+            os.remove(wants_symlink_path)
+            os.symlink(systemd_file_path, wants_symlink_path)
+
+    def remove_unit_files(self):
+        if not os.path.exists(self.__unit_directory):
+            return
+
+        for unit_name in self.unit_names(self.__unit_directory):
+            try:
+                os.remove(os.path.join(self.__base_systemd, unit_name))
+            except FileNotFoundError:
```

Comment:
it seems that if, for some reason, the file was not found, it'll just silently ignore, would not it be more useful catch an exception in the caller?

Reply:
Catching this in the caller would mean exiting the method, which we don't want. This is existing behavior that I don't want to change in this PR.

Reply:
That would mean the `Install` class would have to inspect the wants directory for unit names and then pass them to this method. Knowing how unit files are organized in a wants directory and checking for reserved unit names seem like details that should belong to the `Systemd` class.

Reply:
right, but we could set a flag to file not found and blow up at the very end when we finished the loop,
this would allow us to remove all files and signal back to a caller that something went wrong

Reply:
It's not necessarily an error if we don't find the unit file. We clean up unit files as a sanity check before activating the new set, in which case we don't actually expect to find any.